rust/lance-index/src/scalar/ngram.rs

```diff
@@ -0,0 +1,456 @@
+                )
+                .buffer_unordered(self.io_parallelism)
+                .try_collect::<Vec<_>>()
+                .await?;
+                let list_refs = posting_lists.iter().map(|list| list.as_ref());
+                let row_ids = NGramPostingList::intersect(list_refs);
+                Ok(SearchResult::AtMost(RowIdTreeMap::from(row_ids)))
+            }
+        }
+    }
+
+    fn can_answer_exact(&self, _: &dyn AnyQuery) -> bool {
+        false
+    }
+
+    /// Load the scalar index from storage
+    async fn load(store: Arc<dyn IndexStore>) -> Result<Arc<Self>>
+    where
+        Self: Sized,
+    {
+        Ok(Arc::new(Self::from_store(store.as_ref()).await?))
+    }
+
+    /// Remap the row ids, creating a new remapped version of this index in `dest_store`
+    async fn remap(
+        &self,
+        _mapping: &HashMap<u64, Option<u64>>,
+        _dest_store: &dyn IndexStore,
+    ) -> Result<()> {
+        todo!()
+    }
+
+    /// Add the new data into the index, creating an updated version of the index in `dest_store`
+    async fn update(
+        &self,
+        _new_data: SendableRecordBatchStream,
+        _dest_store: &dyn IndexStore,
+    ) -> Result<()> {
+        todo!()
+    }
+}
+
+pub struct NGramIndexBuilder {
+    tokenizer: TextAnalyzer,
+    tokens_map: HashMap<String, u32>,
+    tokens_counter: u32,
+    bitmaps: Vec<RoaringTreemap>,
+}
+
+impl Default for NGramIndexBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl NGramIndexBuilder {
+    pub fn new() -> Self {
+        let tokenizer = NGRAM_TOKENIZER.clone();
+        Self {
+            tokenizer,
+            // Default capacity loosely based on case insensitive ascii ngrams with punctuation stripped
+            tokens_map: HashMap::with_capacity(4 * 26 * 26 * 26),
+            bitmaps: Vec::with_capacity(4 * 26 * 26 * 26),
+            tokens_counter: 0,
+        }
+    }
+
+    fn validate_schema(schema: &Schema) -> Result<()> {
+        if schema.fields().len() != 2 {
+            return Err(Error::InvalidInput {
+                source: "Ngram index schema must have exactly two fields".into(),
+                location: location!(),
+            });
+        }
+        if *schema.field(0).data_type() != DataType::Utf8 {
+            return Err(Error::InvalidInput {
+                source: "First field in ngram index schema must be of type Utf8".into(),
+                location: location!(),
+            });
+        }
+        if *schema.field(1).data_type() != DataType::UInt64 {
+            return Err(Error::InvalidInput {
+                source: "Second field in ngram index schema must be of type UInt64".into(),
+                location: location!(),
+            });
+        }
+        Ok(())
+    }
+
+    fn process_batch(&mut self, batch: &RecordBatch) {
+        let text_col = batch.column(0).as_string::<i32>();
+        let row_id_col = batch.column(1).as_primitive::<UInt64Type>();
```

Comment:
Should do we assert that there are zero nulls in both columns?

Reply:
Good question.  Bad things.  I will fix this and add tests.